### PR TITLE
WIP commit for fixing file sending on windows

### PR DESCRIPTION
**Pull request by GrayHatter**
This change is [Reviewable](https://reviewable.io/reviews/utox/utox/622)

### CODE REVIEW

File: src/file_transfers.c [hunk 1]
```diff
@@ -868,7 +868,7 @@ uint32_t ft_send_file(Tox *tox, uint32_t friend_number, FILE *file, uint8_t *pat
     const uint8_t *name;
     size_t name_length = 0;
     name = path + path_length;
-    while (*--name != '/') {
+    while (*--name != '\\') {
```
Comment (cebe) [inline src/file_transfers.c]: wouln't this break linux?

Comment (GrayHatter) [inline src/file_transfers.c]: Yup, this needs to be fixed inside windows. I'm officially making the decision that uTox core will never support broken paths. I'll write the fix on the windows side.

Comment (robinlinden) [inline src/file_transfers.c]: @GrayHatter https://github.com/uTox/uTox/blob/develop/src/file_transfers.c#L205

**Comment by GrayHatter** [comment]
closes #600 
closes #601

**Comment by GrayHatter** [comment]
@uTox/devs signoff needed

**Comment by cebe** [comment]
Reviewed 2 of 2 files at r2.
Review status: all files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[src/file_transfers.c, line 1076 at r2](https://reviewable.io:443/reviews/utox/utox/622#-KawbokRFViWbmehyEgJ:-KawbokRFViWbmehyEgK:bbpd9n6) ([raw file](https://github.com/utox/utox/blob/b0de0de9583b2c7313e47e70249b0cb3b3bdd04b/src/file_transfers.c#L1076)):*
> ```C
>         debug_error("FileTransfer:\tUnable to grab a file to start the write friend %u, file %u.",
>                     friend_number, file_number);
>     }
> ```

return missing, you do not want to continue if ft is null.

---

*[src/windows/dnd.c, line 110 at r2](https://reviewable.io:443/reviews/utox/utox/622#-KawcvJLBcJLOB4vJxwc:-KawcvJM0XvZqdkbEql-:b5hfbmt) ([raw file](https://github.com/utox/utox/blob/b0de0de9583b2c7313e47e70249b0cb3b3bdd04b/src/windows/dnd.c#L110)):*
> ```C
>             // Convert windows paths to Unix style;
>             uint8_t *p = path;
>             while (*++p != 0) {
> ```

this will start at the second character in a string. is `\\` a valid beginning of a path in windows? I remember network shares did something like that but I may be wrong.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/utox/utox/622)*

**Comment by robinlinden** [comment]
Reviewed 2 of 2 files at r2.
Review status: all files reviewed at latest revision, 8 unresolved discussions, some commit checks failed.

---

*[src/windows/dnd.c, line 86 at r2](https://reviewable.io:443/reviews/utox/utox/622#-KawkzwGIFsq3VDSMRcu:-KawkzwGIFsq3VDSMRcv:bslwcw6) ([raw file](https://github.com/utox/utox/blob/b0de0de9583b2c7313e47e70249b0cb3b3bdd04b/src/windows/dnd.c#L86)):*
> ```C
>         debug_info("%u files dropped\n", count);
> 
>         for(int i = 0; i < count; i++) {
> ```

Missing space. Should be `for (`

---

*[src/windows/dnd.c, line 88 at r2](https://reviewable.io:443/reviews/utox/utox/622#-KawoN0kdd3sF4-34NXc:-KawoN0l4AmxDPEH8wLL:b-ozf6r9) ([raw file](https://github.com/utox/utox/blob/b0de0de9583b2c7313e47e70249b0cb3b3bdd04b/src/windows/dnd.c#L88)):*
> ```C
>         for(int i = 0; i < count; i++) {
>             debug_notice("WINDND:\tSending file number %i\n", i);
>             UTOX_MSG_FT *msg = calloc(count, sizeof(UTOX_MSG_FT));
> ```

So you loop from 0 to `count` times and allocate `count` objects of size `UTOX_MSG_FT` each iteration of the loop? This doesn't sound right to me..

You probably want to allocate one file_transfer at a time. Or you could keep this allocation, move it to outside of the loop, and access it like `msg[0..count]`.

---

*[src/windows/dnd.c, line 97 at r2](https://reviewable.io:443/reviews/utox/utox/622#-KawpUtv-EInNWazE94Z:-KawpUtv-EInNWazE94_:brsh5lp) ([raw file](https://github.com/utox/utox/blob/b0de0de9583b2c7313e47e70249b0cb3b3bdd04b/src/windows/dnd.c#L97)):*
> ```C
>             if (!path) {
>                 debug_error("WINDND:\tUnable to alloc for UTOX_MSG_FT\n");
>                 return 0;
> ```

Free msg before returning here.

---

*[src/windows/dnd.c, line 102 at r2](https://reviewable.io:443/reviews/utox/utox/622#-KawrVllkD_b_AbF3Pws:-KawrVlmq4eXNhlPHwRn:bf4221l) ([raw file](https://github.com/utox/utox/blob/b0de0de9583b2c7313e47e70249b0cb3b3bdd04b/src/windows/dnd.c#L102)):*
> ```C
>             DragQueryFile(h, i, path, UTOX_FILE_NAME_LENGTH);
> 
>             msg->file = fopen(path, "rb");
> ```

Drop the b. 

---

*[src/windows/dnd.c, line 105 at r2](https://reviewable.io:443/reviews/utox/utox/622#-KawrbdD2LCmnSMY_cwH:-KawrbdD2LCmnSMY_cwI:b-uys62s) ([raw file](https://github.com/utox/utox/blob/b0de0de9583b2c7313e47e70249b0cb3b3bdd04b/src/windows/dnd.c#L105)):*
> ```C
>             if (!msg->file) {
>                 debug_error("WINDND:\tUnable to read file %s\n", path);
>                 return 0;
> ```

Free everything first.

---

*[src/windows/dnd.c, line 110 at r2](https://reviewable.io:443/reviews/utox/utox/622#-KawcvJLBcJLOB4vJxwc:-KawlMDtcpDOw-tJv9zU:b-v5jecl) ([raw file](https://github.com/utox/utox/blob/b0de0de9583b2c7313e47e70249b0cb3b3bdd04b/src/windows/dnd.c#L110)):*
<details><summary><i>Previously, cebe (Carsten Brandt) wrote…</i></summary><blockquote>

this will start at the second character in a string. is `\\` a valid beginning of a path in windows? I remember network shares did something like that but I may be wrong.
</blockquote></details>

You're right, CeBe. `\\network-drive\path\as\usual`

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/utox/utox/622)*

**Comment by GrayHatter** [comment]
Review status: 1 of 2 files reviewed at latest revision, 8 unresolved discussions.

---

*[src/file_transfers.c, line 871 at r1](https://reviewable.io:443/reviews/utox/utox/622#-KaYN8zVMuCaZoCgz4IN-r1-871:-KaxW_PGA5X0g0zbkGfH:b-aetdfk) ([raw file](https://github.com/utox/utox/blob/e11672a16d2ca0c0db4b9eb963168fd98aee7404/src/file_transfers.c#L871)):*
<details><summary><i>Previously, robinlinden (Robin Lindén) wrote…</i></summary><blockquote>

@GrayHatter https://github.com/uTox/uTox/blob/develop/src/file_transfers.c#L205
</blockquote></details>

this was fixed

---

*[src/file_transfers.c, line 1076 at r2](https://reviewable.io:443/reviews/utox/utox/622#-KawbokRFViWbmehyEgJ:-KaxX16H8FnW9n9vmjCU:b-896fix) ([raw file](https://github.com/utox/utox/blob/b0de0de9583b2c7313e47e70249b0cb3b3bdd04b/src/file_transfers.c#L1076)):*
<details><summary><i>Previously, cebe (Carsten Brandt) wrote…</i></summary><blockquote>

return missing, you do not want to continue if ft is null.
</blockquote></details>

Done.

---

*[src/windows/dnd.c, line 86 at r2](https://reviewable.io:443/reviews/utox/utox/622#-KawkzwGIFsq3VDSMRcu:-KaxX4uN3BMvpo1nfAiN:b-896fix) ([raw file](https://github.com/utox/utox/blob/b0de0de9583b2c7313e47e70249b0cb3b3bdd04b/src/windows/dnd.c#L86)):*
<details><summary><i>Previously, robinlinden (Robin Lindén) wrote…</i></summary><blockquote>

Missing space. Should be `for (`
</blockquote></details>

Done.

---

*[src/windows/dnd.c, line 88 at r2](https://reviewable.io:443/reviews/utox/utox/622#-KawoN0kdd3sF4-34NXc:-KaxXXxzAbxw3BEZ5xFy:b-896fix) ([raw file](https://github.com/utox/utox/blob/b0de0de9583b2c7313e47e70249b0cb3b3bdd04b/src/windows/dnd.c#L88)):*
<details><summary><i>Previously, robinlinden (Robin Lindén) wrote…</i></summary><blockquote>

So you loop from 0 to `count` times and allocate `count` objects of size `UTOX_MSG_FT` each iteration of the loop? This doesn't sound right to me..

You probably want to allocate one file_transfer at a time. Or you could keep this allocation, move it to outside of the loop, and access it like `msg[0..count]`.
</blockquote></details>

Done.

---

*[src/windows/dnd.c, line 97 at r2](https://reviewable.io:443/reviews/utox/utox/622#-KawpUtv-EInNWazE94Z:-KaxXYLXoCxM8s2dNH9A:b-896fix) ([raw file](https://github.com/utox/utox/blob/b0de0de9583b2c7313e47e70249b0cb3b3bdd04b/src/windows/dnd.c#L97)):*
<details><summary><i>Previously, robinlinden (Robin Lindén) wrote…</i></summary><blockquote>

Free msg before returning here.
</blockquote></details>

Done.

---

*[src/windows/dnd.c, line 102 at r2](https://reviewable.io:443/reviews/utox/utox/622#-KawrVllkD_b_AbF3Pws:-KaxX_FhqGSxBkxh-UQW:bm46vll) ([raw file](https://github.com/utox/utox/blob/b0de0de9583b2c7313e47e70249b0cb3b3bdd04b/src/windows/dnd.c#L102)):*
<details><summary><i>Previously, robinlinden (Robin Lindén) wrote…</i></summary><blockquote>

Drop the b. 
</blockquote></details>

this is windows, it's needed to prevent mangling of line endings

---

*[src/windows/dnd.c, line 105 at r2](https://reviewable.io:443/reviews/utox/utox/622#-KawrbdD2LCmnSMY_cwH:-KaxXedeHuEDgrj4gQvE:b-896fix) ([raw file](https://github.com/utox/utox/blob/b0de0de9583b2c7313e47e70249b0cb3b3bdd04b/src/windows/dnd.c#L105)):*
<details><summary><i>Previously, robinlinden (Robin Lindén) wrote…</i></summary><blockquote>

Free everything first.
</blockquote></details>

Done.

---

*[src/windows/dnd.c, line 110 at r2](https://reviewable.io:443/reviews/utox/utox/622#-KawcvJLBcJLOB4vJxwc:-KaxXzk4xKa_IDUhh6cI:bvc02qt) ([raw file](https://github.com/utox/utox/blob/b0de0de9583b2c7313e47e70249b0cb3b3bdd04b/src/windows/dnd.c#L110)):*
<details><summary><i>Previously, robinlinden (Robin Lindén) wrote…</i></summary><blockquote>

You're right, CeBe. `\\network-drive\path\as\usual`
</blockquote></details>

This is acceptable, because we're sending the open file as well, this is only so the file transfer code is able to determine the name of the FT by stopping at the first `/`

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/utox/utox/622)*

**Comment by robinlinden** [comment]
:lgtm:

---

Reviewed 1 of 2 files at r3, 1 of 1 files at r4.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[src/windows/dnd.c, line 102 at r2](https://reviewable.io:443/reviews/utox/utox/622#-KawrVllkD_b_AbF3Pws:-Kax_Cgi3pWKW-fYt0Yd:biz95xy) ([raw file](https://github.com/utox/utox/blob/b0de0de9583b2c7313e47e70249b0cb3b3bdd04b/src/windows/dnd.c#L102)):*
<details><summary><i>Previously, GrayHatter (Gregory Mullen) wrote…</i></summary><blockquote>

this is windows, it's needed to prevent mangling of line endings
</blockquote></details>

Huh. You're right. Why can't they just follow the standard? :<

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/utox/utox/622#-:-Kax_S6wiJaVe9mgoROL:bnfp4nl)*

**Comment by cebe** [comment]
Reviewed 1 of 2 files at r3, 1 of 1 files at r4.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[src/windows/dnd.c, line 110 at r2](https://reviewable.io:443/reviews/utox/utox/622#-KawcvJLBcJLOB4vJxwc:-Kax_El0s01U9sU83Kjl:b-zibenp) ([raw file](https://github.com/utox/utox/blob/b0de0de9583b2c7313e47e70249b0cb3b3bdd04b/src/windows/dnd.c#L110)):*
<details><summary><i>Previously, GrayHatter (Gregory Mullen) wrote…</i></summary><blockquote>

This is acceptable, because we're sending the open file as well, this is only so the file transfer code is able to determine the name of the FT by stopping at the first `/`
</blockquote></details>

if this is only needed for the file name it would be more sane to just cut it here instead of sending a broken path out.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/utox/utox/622)*

**Comment by GrayHatter** [comment]
Review status: 0 of 2 files reviewed at latest revision, 1 unresolved discussion.

---

*[src/windows/dnd.c, line 110 at r2](https://reviewable.io:443/reviews/utox/utox/622#-KawcvJLBcJLOB4vJxwc:-KaxcHffdFMVVjGF3One:b-k38g2d) ([raw file](https://github.com/utox/utox/blob/b0de0de9583b2c7313e47e70249b0cb3b3bdd04b/src/windows/dnd.c#L110)):*
<details><summary><i>Previously, cebe (Carsten Brandt) wrote…</i></summary><blockquote>

if this is only needed for the file name it would be more sane to just cut it here instead of sending a broken path out.
</blockquote></details>

changed and moved to the file transfer code

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/utox/utox/622)*

**Comment by robinlinden** [comment]
:lgtm_strong:

---

Reviewed 1 of 2 files at r5, 1 of 1 files at r6, 1 of 1 files at r7.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/utox/utox/622#-:-KaxeM1GS9Usf-7Q1_dn:b-jbsf8z)*

**Comment by cebe** [comment]
Reviewed 2 of 2 files at r5.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[src/file_transfers.c, line 210 at r6](https://reviewable.io:443/reviews/utox/utox/622#-Kaxdon84l2LUNiVjR1j:-Kaxdon84l2LUNiVjR1k:b-849lta) ([raw file](https://github.com/utox/utox/blob/c3665d90c9cf034320764e19090a3a08604f9ca7/src/file_transfers.c#L210)):*
> ```C
>     ft->name_length = 0;
>     uint8_t *p = ft->path + strlen((char *)ft->path);
>     while (*--p != '/' || *p != '\\') {
> ```

NOO!

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/utox/utox/622)*

**Comment by cebe** [comment]
:lgtm:

---

Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/utox/utox/622#-:-Kaxq8EIJjqIoezgea1A:bnfp4nl)*

**Comment by cebe** [comment]
Reviewed 1 of 1 files at r7.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/utox/utox/622)*

**Pull request merged**